sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -212,9 +210,18 @@ class StreamExecution(
       populateStartOffsets()
       logDebug(s"Stream running from $committedOffsets to $availableOffsets")
       while (isActive) {
+        val batchStartTimeMs = System.currentTimeMillis()
```

Comment:
My original design is as follows:

```Scala
trait Trigger {

  def triggerExecution(): TriggerExecution
}

trait TriggerExecution {

  // batchRunner will run a batch and run if we should terminate the execution
  def execute(batchRunner: () => Boolean): Unit
}

private[sql] class ProcessingTime(intervalMs: Long) extends Trigger {
  override def triggerExecution(): TriggerExecution = new ProcessingTriggerExecution(intervalMs)
}

private[sql] class ProcessingTriggerExecution(intervalMs: Long) extends TriggerExecution {
  override def execute(batchRunner: () => Boolean): Unit = {
    while (true) {
      // Codes that waits for next multiple of intervalMs.
      if (batchRunner()) {
        return
      }
    }
  }
}
```
However, it requires exposing `Trigger` and `TriggerExecution` to the user and seems complicated.